src/samplebuffer.h

```diff
@@ -0,0 +1,73 @@
+#ifndef SAMPLEBUFFER_H
+#define SAMPLEBUFFER_H
+
+#include "util/types.h" // CSAMPLE
+
+#include <algorithm> // std::swap
+
+// A sample buffer with properly aligned memory to enable SSE optimizations.
+// The public interface closely follows that of std::vector.
+//
+// No resize operation is provided intentionally for maximum efficiency!
+// If the size of an existing sample buffer needs to be altered after
+// construction this can simply be achieved by swapping the contents with
+// a temporary sample buffer that has been constructed with the desired
+// size:
+//
+//     SampleBuffer sampleBuffer(oldSize);
+//     ...
+//     SampleBuffer tempBuffer(newSize)
+//     ... copy data from sampleBuffer to tempBuffer...
+//     sampleBuffer.swap(sampleBuffer);
+//
+// The local variable tempBuffer can be omitted if no data needs to be
+// copied from the existing sampleBuffer:
+//
+//     SampleBuffer sampleBuffer(oldSize);
+//     ...
+//     SampleBuffer(newSize).swap(sampleBuffer);
+//
+// After construction the content of the buffer is uninitialized.
+class SampleBuffer {
+    Q_DISABLE_COPY(SampleBuffer);
+
+  public:
+    typedef CSAMPLE value_type;
+    typedef int size_type;
```

Comment:
https://software.intel.com/en-us/articles/about-size-t-and-ptrdiff-t

According to Intel size_t (unsigned) and ptrdiff_t (signed) are the natural choice for "...loop counters, array indexing, size storage and address arithmetic."

Another statement is about portability between 32-/64-bit: "...where use of size_t and ptrdiff_t types provides high performance, possibility to operate large data sizes and portability between different platforms."

My conclusion: ptrdiff_t (signed) and size_t (unsigned) should be used for loop counters, array indexing, size storage and address arithmetic to achieve maximum portability and performance. If you don't want to deal with signed/unsigned issues ptrdiff_t would be the natural choice instead of int. Writing "ptrdiff_t" everywhere in the code is a bit awkward and a custom typedef might help here.